python: SparkAPI python wrapper

erizocosmico opened the pull request:
First draft of the SparkAPI python wrapper.

### What's missing?

* [x] Tests! Although it has been tested IRL, it might be good to add some tests. But when the API is stable (or as stable as it can be) they can be added. After all, it's a very thin layer.
* [x] Auto-add and manage the SparkAPI jar. **Can't be done**
*  README. Should this package have it's own README or should all its contents be on the main README? Being done in #43 
* [x] travis jobs
* [x] pypi release on tag with travis
* [x] Update python notebook Dockerfile demo **(blocked until #38 is merged)**

### Example

```python
from sourced.spark import API as SparkAPI
from pyspark.sql import SparkSession

spark = SparkSession.builder.appName("test").master("local[*]").getOrCreate()
api = SparkAPI(spark, '/path/to/siva/files')
api.repositories.filter("url = 'yadayada.foo/url.git'").references.head_ref.commits.show()
spark.stop()
```

* Install the python wrapper doing `python setup.py install` or `python setup.py install --user` or add the `python` folder of the `spark-api` repo to your `PYTHONPATH` (`export PYTHONPATH="$PYTHONPATH:/path/to/python/folder"`).
* Run with `spark-submit --jars path/to/spark-api-jar /path/to/this/example.py`

  Can also be run from pyspark shell (adding the `--jar /path/to/spark-api-jar` when starting the shell). In the case of the pyspark shell you don't need to create the spark session, it's already available as `spark`.

### Design decisions

Because with Py4J and Python we lose the Scala implicits we lose the ability to automatically add our methods to `DataFrame`. Because of that, we have a `SourcedDataFrame` which is a subclass of `pyspark.sql.DataFrame`. But because the methods return new `DataFrame`s whenever we use a method of `DataFrame` we are stuck with a plain `DataFrame`.

For example, doing the following would not be valid:

```python
api.repositories.filter("some condition").references
```

Because `DataFrame` does not have a `references` method or property.

To overcome this, `SourcedDataFrame` overrides the methods of `DataFrame` that return a new `DataFrame` and transforms them to the corresponding class. For example, if we have an instance of `RepositoriesDataFrame`, any method that returns a new `DataFrame` now will return an instance of `RepositoriesDataFrame` instead, and so on.

To do so and not care much about API changes in `spark.sql.DataFrame` it's done using a decorator that wraps the `DataFrame` with the actual concrete dataframe it's being used and all the parameters are passed with `*args` and `**kwargs`. So, if the arguments of these functions change we don't care, our overridden methods are acting just as a proxy between the user and the actual `pyspark.sql.DataFrame` methods and if they change we don't have to change anything. The only API changes we'd care would be added methods or removed methods.
While this is a very tedious task having to write the overridden method for each method, it's really straightforward, as it's always the same template:

```python
@custom_df_instance
def METHOD_NAME(self, *args, **kwargs):
    return DataFrame.METHOD_NAME(self, *args, **kwargs)
```

As a result of this, we can use any `pyspark.sql.DataFrame` method we want without losing the methods added by our subclasses.

The rest of the API is very straightforward, as it's just wrapping scala functionality.

bzz commented:
Quick question: do you think `Example` section above could be updated to include an instructions on how to start `pyspark` shell with SparkAPI dependency?

erizocosmico commented:
`python setup.py install` will install the dependency as a python library. Then it can be used from elsewhere.

I'll update the snippet so it's just copy and pasteable.

bzz commented:
@erizocosmico indeed, that would work only in local mode, when Spark is running on a single machine.
In case of cluster - that would require users (us) to manually installing that dependency on every Worker, which is quite tiresome.

Apache Spark has [tooling to avoid that](https://spark.apache.org/docs/latest/submitting-applications.html#advanced-dependency-management) i.e `spark-submit --jars`, in case of Scala and `--py-files` for python, that would distribute the dependencies with the job all over cluster for you.

What I meant is - it would be really nice if we could make sure SparkAPI will follow this practice, so that there is a simple command like `./spark-shell <...>` and then `from sourced.spark import API as SparkAPI` works without any pre-requests (except for having Apache Spark, of course).

One more question  - do you think we could add `.parseUASTs()` and `.classifyLanguage()` to the wrapper as well? This would put in place a templates, that could be filled in with actual impls in subsequent PRs (i.e #41 could include that, if this is merge first).

erizocosmico commented:
@bzz yes, this only works right now on local mode. I've been trying to autoregister the jar from inside pyspark but it's been proving impossible to do so far. Since we have to add the `--py-files` anyway for distributing the python wrapper across the workers we might as well distribute the jar file via a command. I will make that change in another pull request after this one.

Is it worth adding the `parseUASTs` and `classifyLanguage` if it's not gonna have implementation right now? I mean, since it's not implemented yet in Scala (or wasn't when I started this) there's nothing to call yet and when it's implemented in scala changes would be required here as well. Is there any reason I'm missing for implementing this two empty methods right now? Also, these methods only make sense in the `BlobsDataFrame`, right?

bzz commented:
Having `wget spark-api.zip<or something else>.zip` that is built as a part of SparkAPI standard release process and uploaded on Github sounds as a good first step. This can be automated by CI i.e on-tag.

This would allow us to still have both: a beginner-friendly on-boarding process (few commands) and cover both, local and cluster modes with (until it's published on Maven central)
```
wget 
pyspark  --repositories "https://jitpack.io" \
              --packages "tech.sourced:spark-api:0.1.0-SNAPSHOT" \
              --py-files ./spark-api.zip
``` 

In future, this could be replaced with `pip install <src-d-spark-api>` which is even more reasonable for Python ecosystem and would not change how we run things.

What do you think?

bzz commented:
@erizocosmico 
> Is it worth adding the parseUASTs and classifyLanguage if it's not gonna have implementation right now? 
> Also, these methods only make sense in the BlobsDataFrame, right?

I believe yes and yes :) In my opinion, it's worth because #42 is ready to be merged, and this seems to be almost ready, so if this get's rebased - it could include `classifyLanguages()` impl and if it get's merged then -  #41 could take care of filling the template for UAST.

But it's totally up to you.

BTW, from the list of missing things in description - https://github.com/src-d/spark-api/pull/43 is taking care of initial work of showing simple example.

erizocosmico commented:
@bzz sure!

erizocosmico commented:
Just updated the PR with the travis jobs to run the tests of the python wrapper. It's a pity we have to duplicate so much code, but in `matrix.include` you can only specify a single version :(

erizocosmico commented:
Updated with tag autorelease in travis (example can be seen here https://test.pypi.org/project/test-sourced-spark-api/#history)

erizocosmico commented:
Done!

Just updated the Basic Example python notebook to use the python wrapper. Instead of using a zip I just installed it locally with pip.

Also converted the example code to snake case, which is the python naming convention for everything but classes (and which pyspark seems to deliberately ignore).

erizocosmico commented:
With this last commit tests will fail until #45 is merged

erizocosmico commented:
Rebased, CI should be ok now

erizocosmico commented:
Ready to merge when CI decides to finish

bzz commented:
CI is green now! 

While running it locally with

```
git fetch upstream pull/39/head:feature/python-wrapper
./sbt clean assembly
docker build -t spark-api-jupyter .
docker run -d -p 8888:8888 -v "$(pwd)/src/test/resources/siva-files:/repositories" spark-api-jupyter
```

And in web UI, after running the first paragraph have
```
Py4JError: An error occurred while calling o33.registerUDFs. Trace:
py4j.Py4JException: Method registerUDFs([]) does not exist
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:318)
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:326)
	at py4j.Gateway.invoke(Gateway.java:272)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)

```

Full log in details
<details>

```
---------------------------------------------------------------------------
Py4JError                                 Traceback (most recent call last)
<ipython-input-1-6341628568e7> in <module>()
      4 spark = SparkSession.builder.master("local").appName("Examples").getOrCreate()
      5 
----> 6 api = SparkAPI(spark, "/repositories")

/opt/python-spark-api/sourced/spark/api.py in __init__(self, session, repos_path)
     24         self.__api = self.__jvm.tech.sourced.api.SparkAPI(self.__jsparkSession)
     25         self.__api.setRepositoriesPath(repos_path)
---> 26         self.__api.registerUDFs()
     27         self.__implicits = getattr(getattr(self.__jvm.tech.sourced.api, 'package$'), 'MODULE$')
     28 

/usr/local/spark/python/lib/py4j-0.10.4-src.zip/py4j/java_gateway.py in __call__(self, *args)
   1131         answer = self.gateway_client.send_command(command)
   1132         return_value = get_return_value(
-> 1133             answer, self.gateway_client, self.target_id, self.name)
   1134 
   1135         for temp_arg in temp_args:

/usr/local/spark/python/pyspark/sql/utils.py in deco(*a, **kw)
     61     def deco(*a, **kw):
     62         try:
---> 63             return f(*a, **kw)
     64         except py4j.protocol.Py4JJavaError as e:
     65             s = e.java_exception.toString()

/usr/local/spark/python/lib/py4j-0.10.4-src.zip/py4j/protocol.py in get_return_value(answer, gateway_client, target_id, name)
    321                 raise Py4JError(
    322                     "An error occurred while calling {0}{1}{2}. Trace:\n{3}\n".
--> 323                     format(target_id, ".", name, value))
    324         else:
    325             raise Py4JError(

Py4JError: An error occurred while calling o33.registerUDFs. Trace:
py4j.Py4JException: Method registerUDFs([]) does not exist
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:318)
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:326)
	at py4j.Gateway.invoke(Gateway.java:272)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)
```
</details>

Could you please help me to understand - did I miss something or is that an expected result right now?

erizocosmico commented:
@bzz was indeed a bug. What I don't know is why CI was green :/. Should be fixed now.

bzz commented:
@erizocosmico awesome, thanks! Let me try it one more time

bzz commented:
One question - what would be equivalent in PySpark of this [Scala API call sequence](https://github.com/src-d/spark-api/blob/master/src/test/scala/tech/sourced/api/DefaultSourceSpec.scala#L64) that we use in tests and is covering one of the use-cases:

```scala
    val filesDf = SparkAPI(spark, resourcePath)
      .getRepositories.filter($"id" === "github.com/mawag/faq-xiyoulinux")
      .getReferences.filter($"name".equalTo("refs/heads/HEAD"))
      .getFiles
      .select("repository_id", "name", "path", "commit_hash", "file_hash", "content")
```
?

erizocosmico commented:
```python
files_df = SparkAPI(spark, resourcePath).repositories\
    .filter("id = 'github.com/mawag/faq-xiyoulinux'")\ # or .filter(df.id == 'github.com/mawag/faq-xiyoulinux') if you have a repos df reference
    .references.head_ref.blobs\
    .select("repository_id", "name", "path", "commit_hash", "file_hash", "content")
```

Instead of files it's called blobs, because I started implementing this before merging the BlobIterator and I figured it would be renamed to `blob`, we can change to files if you prefer it to have more similar APIs on both sides.

Another difference is the `head_ref` which is a shortcut for filtering only heads. We could add these shortcuts on the scala api as well.

bzz commented:
@erizocosmico thank you for explaining! 

+1 for `blobs` -> `files` in this PR and add `.headRef` and `.masterRef` to Scala API in another

bzz commented:
Just FYI, for me `api.repositories.references.head_ref.blobs` in Jupyter results in 

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-12-d948dd0bc6b7> in <module>()
----> 1 api.repositories.references.head_ref.blobs

/usr/local/spark/python/pyspark/sql/dataframe.py in __getattr__(self, name)
   1018         if name not in self.columns:
   1019             raise AttributeError(
-> 1020                 "'%s' object has no attribute '%s'" % (self.__class__.__name__, name))
   1021         jc = self._jdf.apply(name)
   1022         return Column(jc)

AttributeError: 'ReferencesDataFrame' object has no attribute 'blobs'
```

erizocosmico commented:
Yes, you can only get the blobs from a `CommitsDataFrame`, just like with the Scala API (which checks for a "files" column that only after joinin with commits is available)

You need to `head_ref.commits.blobs` in this case `head_ref.commits.files` when I upload the change

bzz commented:
@erizocosmico not sure I understood you, could you please elaborate?

What I was saying is - this already works in Scala
```scala
val filesDf = SparkAPI(spark, resourcePath)
      .getRepositories.filter($"id" === "github.com/mawag/faq-xiyoulinux")
      .getReferences.filter($"name".equalTo("refs/heads/HEAD"))
      .getFiles
      .select("repository_id", "name", "path", "commit_hash", "file_hash", "content")
``` 
and that would be nice to have same ability in Python. 

Or am I missing something here?

erizocosmico commented:
Sorry, nvm, thought the files were only accessible after asking for the commits. Will update this so it has parity with the Scala API.

erizocosmico commented:
Done @bzz

bzz commented:
Sounds great, thanks!
Let me try again.
 
It all shows that we kind-of need to start thinking about automating this parity testing thing somehow...

bzz commented:
Meanwhile, just noticed that CI fails for some reason now :/

erizocosmico commented:
@bzz the race condition fixed in #48

bzz commented:
@erizocosmico I have tried

```
api.repositories.references.head_ref.files.classify_languages().select("id", "path", "lang")
```

and got
```
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-7-db2d73d96ab5> in <module>()
----> 1 api.repositories.references.head_ref.files.classify_languages().select("id", "path", "lang")

/opt/python-spark-api/sourced/spark/api.py in classify_languages(self)
    350         its row.
    351         """
--> 352         raise NotImplementedError("classify_languages not yet implemented")
    353 
    354 

NotImplementedError: classify_languages not yet implemented
```
 
Am I doing something wrong? AFAIK it was implemented in Scala in now merged #42

erizocosmico commented:
Oh, sure, it wasn't updated when the PR got merged. Updating

erizocosmico commented:
@bzz fixed. While implementing this I've found a bug on the `checkCols` function that we haven't noticed yet because `classifyLanguages` is not used in any test and we've never used more than one column before.

bzz commented:
@erizocosmico thanks, let me try again! 

The bug in `checkCols` is fixed under https://github.com/src-d/spark-api/pull/41/commits/617ede92adacc30dd0d00f8e39af859549e5cf90

bzz commented:
On local build of enry on osx, it's impossible to build working Docker container

```
api.repositories.references.head_ref.files.classify_languages().select("repository_id", "path", "lang").show(20, False)
```

results in

```
Py4JJavaError: An error occurred while calling o282.showString.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 29.0 failed 1 times, most recent failure: Lost task 0.0 in stage 29.0 (TID 427, localhost, executor driver): java.lang.UnsatisfiedLinkError: Unable to load library 'enry': Native library (linux-x86-64/libenry.so) not found in resource path ([file:/opt/jars/spark-api-uber.jar, file:/home/jovyan/spark-api-uber.jar])
	at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:271)
	at com.sun.jna.NativeLibrary.getInstance(NativeLibrary.java:398)
	at com.sun.jna.NativeLibrary.getInstance(NativeLibrary.java:337)
	at tech.sourced.enry.nativelib.EnryLibrary.<clinit>(EnryLibrary.java:13)
	at tech.sourced.enry.Enry.<clinit>(Enry.java:10)
	at tech.sourced.api.udf.ClassifyLanguagesUDF$.getLanguage(ClassifyLanguagesUDF.scala:15)
	at tech.sourced.api.udf.ClassifyLanguagesUDF$$anonfun$1.apply(ClassifyLanguagesUDF.scala:9)
```

:/ 

It's offtopic, but how do you think, is there any chance incorporating some cross-compilation awesomeness in enry-java, without release part of the https://github.com/src-d/enry/compare/master...erizocosmico:feature/maven-publish-jar?expand=1 ?

erizocosmico commented:
@bzz that's what I put in the README, all the operations must be done in the same OS exactly because of that :(. Cross-compilation is not trivial to setup, but can definitely be done. I guess we could do a Dockerfile for cross-compiling in enry-java because asking the user to set that stuff locally is madness.

I will also rebase in a moment this branch and implement the last bits (bblfsh parseUAST) and we'll be ready to go.

erizocosmico commented:
Extract UASTs implemented. CI will fail for now until we submit a fix for the thing @mcarmonaa reported on Slack (guava issue).

Once that's done and this is rebased we're golden.

bzz commented on .travis.yml:
```diff
@@ -24,9 +25,53 @@ before_script:
+    python: 2.7
+
+    install:
+      - sbt assembly
```
let's use a `./sbt` wrapper

erizocosmico commented:
Ready to be merged when CI passes. This branch includes the fix with the reshade for the assembly task

ajnavarro commented on .vscode/tags:
```diff
@@ -0,0 +1,152 @@
+!_TAG_FILE_FORMAT	2	/extended format; --format=1 will not append ;" to lines/
```
what is this file?

erizocosmico replied:
removed

ajnavarro commented on build.sbt:
```diff
@@ -36,6 +36,10 @@ jacoco.reportFormats in jacoco.Config := Seq(
 parallelExecution in Test := false
 logBuffered in Test := false
 
+assemblyShadeRules in assembly := Seq(
```
:+1:

erizocosmico commented:
Yeah, theres an issue with a test in python 3. I'll fixit after lunch
On Wed, 27 Sep 2017 at 13:37, Antonio Navarro Perez <notifications@github.com<mailto:notifications@github.com>> wrote:

@ajnavarro approved this pull request.

LGTM when Travis pass

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/src-d/spark-api/pull/39#pullrequestreview-65510803>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABQFF9XPgAcY0uUdyVlLM4YVFs1L1J3Sks5smjN6gaJpZM4Pa2cD>.

erizocosmico commented:
Fixed @ajnavarro

smacker commented on cmd/sourced/compose/workdir/workdir.go:
```diff
@@ -134,7 +134,7 @@ func UnsetActive() error {
 	}
 
-	_, err = os.Stat(dir)
+	_, err = os.Lstat(dir)
```
could you explain why do you need to use `Lstat` here?

dpordomingo replied:
sure!

from the PR description:
>_`UnsetActive` must work with `Lstat` to check symlink status,
instead of accessing the pointed dir._

From the function docs:
```
// UnsetActive removes symlink for active workdir
```

So let's consider:
```
__active__ -> ~/.sourced/workdirs/this/dir/was/deleted
```

Then `os.Stat(dir)` is going to return `os.ErrNotExist` (because it checks the pointed file, not the symlink itself), so `os.Remove(dir)` is not going to be called, so `__active__` is not going to be deleted.

Using `os.Lstat(dir)` it will check the symlink itself (`__active__`), that exists, so it will be deleted as expected.

smacker replied:
makes sense 👍
